nixos/sslh: update and refactor for RFC42

rnhmjoj opened the pull request:
###### Description of changes

Update the sslh package and rewrite the module for RFC42

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested via `nixosTests.sslh`
- [x] Tested compilation of all packages that depend on this change (`sslh`)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

rnhmjoj committed:
sslh: install sslh-select binary

rnhmjoj committed:
nixos/sslh: refactor for RFC42

rnhmjoj committed:
sslh: 1.22c -> 2.0.0

rnhmjoj commented:
@symphorien This should superseed #225602 in fixing #225454.

infinisil commented:
There are already existing efforts to have a libconfig format generator: https://github.com/NixOS/nixpkgs/pull/246115

rnhmjoj commented:
Wow, that seems way overkill. In comparison, mine is a lot simpler purely Nix implementation of the conversion, it does not handle special ints and arrays, but they could be added in the same way, if needed.

ckiee commented:
There's a Nix-only implementation in #208747 that was iterated on a few times before [this comment](https://github.com/NixOS/nixpkgs/pull/208747#issuecomment-1372545465) which I am taking out of a bit of context:
> It's just that this adds a whole bunch of code that's slow to evaluate, hard to debug, test and maintain. And yes that's half of nixpkgs already but I don't think it's a good idea to make this worse. Others might review and merge this PR, and I won't interfere with that, but personally I'm gonna abstain from this and focus on cleaner code instead.

After a long period of silence @h7x4 came in, reviving that PR with #246115 which also has output validation and a test alongside being written in a strongly-typed language. It should be a lot more robust.

Sorry about the trouble though— I've also been rebasing my PRs until this thing wraps up.

rnhmjoj committed:
nixos/release-notes: mention sslh changes

h7x4 commented on nixos/modules/services/networking/sslh.nix:
```diff
@@ -5,81 +5,131 @@ with lib;
+        };
+
+        options.verbose-connections = mkOption {
+          type = types.enum [ 0 1 2 3 4 ];
```
This is sort of a nit since both of these options work, but I feel like it's more semantically correct to look at this as a discrete range from least verbose to most verbose. Therefore, I think this would be more fitting

```suggestion
          type = types.ints.between 0 4;
```

h7x4 commented on nixos/modules/services/networking/sslh.nix:
```diff
@@ -5,81 +5,131 @@ with lib;
+        freeformType = configFormat.type;
+
+        options.timeout = mkOption {
+          type = types.int;
```
Assuming timeout of `0` is possible as well.

```suggestion
          type = types.ints.unsigned;
```

h7x4 commented on nixos/modules/services/networking/sslh.nix:
```diff
@@ -5,81 +5,131 @@ with lib;
+    };
+
+    listenAddresses = mkOption {
+      type = types.coercedTo types.str singleton (types.listOf types.str);
```
nit, makes it a bit easier to read.

```suggestion
      type = with types; coercedTo str singleton (listOf str);
```

rnhmjoj commented:
Rebased (since https://github.com/NixOS/nixpkgs/pull/246115 has been merged) and added a release note.
Unless there are objections, I'm going to merge this today to get it in 23.11.

h7x4 reviewed (commented):
A few small comments. Some of the comments regard stuff that has just been moved, so feel free to ignore if you don't think it's relevant for the refactor.

rnhmjoj commented:
@h7x4 thank you, I applied you suggestions.
